scripts/jenkins/ardana/ansible/roles/heat_stack/tasks/delete.yml

```diff
@@ -0,0 +1,78 @@
+#
+# (c) Copyright 2018 SUSE LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+---
+
+- name: Ensure stack is not in a pending state before deleting it
+  command: |
+    openstack --os-cloud {{ os_cloud }} stack show {{ heat_stack_name }} \
+      -c stack_status -f value
+  register: stack_status
+  retries: 100
+  until: "'IN_PROGRESS' not in stack_status.stdout"
+  delay: 10
+
+- block:
+    - name: Delete stack '{{ heat_stack_name }}'
+      os_stack:
+        cloud: "{{ os_cloud }}"
+        name: "{{ heat_stack_name }}"
+        state: absent
+        api_timeout: "{{ api_timeout }}"
+      when: "'DELETE_COMPLETE' not in stack_status.stdout"
+  rescue:
+    - name: Force delete stack resources when delete failed
+      command: |
+         openstack --os-cloud {{ os_cloud }} stack resource list --filter \
+           type={{ item.type }} -f value -c physical_resource_id {{ heat_stack_name }} |
+           awk '{print "openstack --os-cloud '{{ os_cloud }}' {{ item.name }} \
+           delete --wait "$1 }' | sh -x || :
+      loop:
+        - name: "server"
+          type: "OS::Nova::Server"
+        - name: "volume"
+          type: "OS::Cinder::Volume"
+        - name: "network trunk"
+          type: "OS::Neutron::Trunk"
+        - name: "port"
+          type: "OS::Neutron::Port"
+        - name: "router"
+          type: "OS::Neutron::Router"
+        - name: "subnet"
+          type: "OS::Neutron::Subnet"
+        - name: "network"
+          type: "Neutron::Net"
+
+    - name: Retry delete stack
+      os_stack:
+        cloud: "{{ os_cloud }}"
+        name: "{{ heat_stack_name }}"
+        state: absent
+        api_timeout: "{{ api_timeout }}"
+      register: stack_status
+      retries: 3
+      until: "stack_status.rc == 0"
```

Comment:
One thing you **absolutely** want to avoid is delete a stack which is in a "DELETE_IN_PROGRESS" state. Doing this could result in un-delete-able stacks, such as the ones that you can see in a DELETE_FAILED status now. Which is why you need to check the stack's state every time before you try to delete it.

Reply:
As we discussed, the status of the stack should never be "DELETE_IN_PROGRESS" when it reaches this task.